.gitmodules

```diff
@@ -130,3 +130,7 @@
 	path = Sming/Libraries/SmingTest
 	url = https://github.com/mikee47/SmingTest
 	ignore = dirty
+[submodule "Sming/Libraries/ModbusMaster"]
+	path = Sming/Libraries/ModbusMaster
+	url = https://github.com/kmihaylov/ModbusMaster.git
```

Comment:
Also, I note in your fork of `ModbusMaster` the method `requestFrom` could probably just be removed?

Reply:
> Do you think the improvements will be pushed back to the original repo at `https://github.com/4-20ma/ModbusMaster`? It would be preferable to use that if possible to avoid divergence.

I forked nomis' since it contains some important modifications according to me. Unfortunately they're not merget into 4-20ma's repo. I think that the original author has abandoned it or something else, but it has no changes for the last 2-3 years... Otherwise it would be nice to stick with the original repo.

> Also, I note in your fork of `ModbusMaster` the method `requestFrom` could probably just be removed?

I'm trying to keep it close to the original, however if it doesn't contradict with Sming's rules for 3rd party libraries, I can remove it?

Reply:
It's your PR! If it's not needed and is going to get removed anyway then may as well.

Reply:
If nomis` fork is going to be the reference you could use that but provide a `.patch` file with any minor amendments for Sming. There's information in the documentation about how to do this, but am working on improvements to the documentation so you might find this more helpful https://smingdev.readthedocs.io/en/latest/contribute/external-sources.html

Reply:
First I thought to make a patch, but I decided to outrun the upcoming changes for the 3rd party libraries, therefore I did a fork. I'm watching #1983.  I'm waiting what changes would you introduce. I do not use MBMaster right now anyway...

Personally I feel more comfortable using authors' repositories, but I'm afraid that lots of changes may be required for ESP32. I couldn't judge this from my perspective...

Reply:
Above I suggested that you can move the original ModbusMaster into a sub-directory, thus creating a Sming Component also called `ModbusMaster`. Another library which does this is `libsodium` which was added quite recently.

The proposals in #1983 would allow the entire Component could be moved into your own repository. It wouldn't require you to make a separate fork of `ModbusMaster` (other than for your own development purposes).

If you wanted to put your sample application into the Component rather than in the Sming `samples` directory, that would work just fine, just a few things to note which I've listed in https://github.com/SmingHub/Sming/issues/1983#issuecomment-563286832.

Reply:
Fine, I forgot to reach smingdev and I left only reading the 'stable' documentation.

I prefer the libsodium's way for putting the library. I moved the library after your directory structure, put ```component.mk``` for the library and now It works, but I have to manually fetch the library before building Sming. I have to issue these commands in this order (obviously Sming doesn't fetch the library for me and I don't know why).

```
cd <sample-path>
make clean
make config-clean
make dist-clean
cd /opt/Sming
git submodule update --init Sming/Libraries/ModbusMaster/ModbusMaster/
cd <sample-path>
make
```

After solving this problem I would also prepare a patch (if needed) and change the submodule path to nomis's fork.

Another question is what extension should I use for the README files, there is .md, .rst... Probably they're related to sphinx?

Reply:
I think with the latest `develop` it got build, but now 
```
Building /opt/Sming/Sming/Libraries/ModbusMaster/samples/general/out/Esp8266/debug/lib/clib-App-1030e4fa9ed7c0bb427712777cc5f58e.a
make[1]: *** No rule to make target '/opt/Sming/Sming/Arch/Esp8266/Components/esp8266/include/c_types.h', needed by 'app/application.o'.  Stop.
/opt/Sming/Sming/project.mk:357: recipe for target 'App-build' failed
make: *** [App-build] Error 2
```

Reply:
See https://sming.readthedocs.io/en/latest/contribute/documentation.html for a primer on documentation.

Reply:
I updated the toolchain and the c_type error is gone. However it seems that Sming doesn't fetch ModbusMaster library automatically for me. I still have to proceed like in https://github.com/SmingHub/Sming/pull/1992#discussion_r356457913

Reply:
Check `Sming$ git status` and remove any extraneous files/directories, e.g. `Sming/Arch/Esp8266/Components/custom_heap/`

Reply:
You need to add to your library's `component.mk` file:

```
COMPONENT_SUBMODULES := ModbusMaster
```

Then do `make ModbusMaster-rebuild all`.